File: codeforces/1900~1999/1902/d.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0, q = 0;
  cin >> n >> q;
  string s;
  cin >> s;

  map<int, map<int, vector<int>>> indices;
  int x = 0, y = 0;
  indices[x][y].push_back(0);
  vector<pair<int,int>> pos;
  pos.emplace_back(x, y);
  for (int i = 0; i < n; ++i) {
    if (s[i] == 'U') {
      y++;
    } else if (s[i] == 'D') {
      y--;
    } else if (s[i] == 'L') {
      x--;
    } else if (s[i] == 'R') {
      x++;
    }

    indices[x][y].push_back(i+1);
    pos.emplace_back(x, y);
  }

  auto query_left = [&](int x, int y, int l) -> bool {
    if (!indices.count(x)) return false;
    if (!indices[x].count(y)) return false;
    vector<int>& idx = indices[x][y];
    return upper_bound(idx.begin(), idx.end(), l-1) != idx.begin();
  };

  auto query_right = [&](int x, int y, int r) -> bool {
    if (!indices.count(x)) return false;
    if (!indices[x].count(y)) return false;
    vector<int>& idx = indices[x][y];
    return lower_bound(idx.begin(), idx.end(), r+1) != idx.end();
  };

  // pos(l-1) + pos(r) - pos(i) = (x, y)
  // => pos(i) = pos(l-1) + pos(r) - (x, y)
  // => find an index i such that i >= l-1 and i <= r
  // in the map
  // x = pos[l-1][0] + pos[r][0] - xi
  // y = pos[l-1][1] + pos[r][1] - yi
  auto query_middle = [&](int xi, int yi, int l, int r) -> bool {
    int x = pos[l-1].first + pos[r].first - xi;
    int y = pos[l-1].second + pos[r].second - yi;
    if (!indices.count(x)) return false;
    if (!indices[x].count(y)) return false;
    vector<int>& idx = indices[x][y];
    auto it = lower_bound(idx.begin(), idx.end(), l-1);
    if (it == idx.end()) return false;
    return *it <= r;
  };

  auto query = [&](int x, int y, int l, int r) -> bool {
    return query_left(x, y, l) || query_right(x, y, r) || query_middle(x, y, l, r);
  };

  for (int i = 0; i < q; ++i) {
    int xi = 0, yi = 0, l = 0, r = 0;
    cin >> xi >> yi >> l >> r;
    bool result = query(xi, yi, l, r);
    if (result) {
      cout << "YES" << endl;
    } else {
      cout << "NO" << endl;
    }
  }
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  solve();
  return 0;
}

File: codeforces/1900~1999/1902/d.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::collections::BTreeMap;

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn read_str() -> String {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().to_string()
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());

    let inputs = read_vec::<i32>();
    let n = inputs[0] as usize;
    let q = inputs[1];
    let s = read_str();
    let b = s.as_bytes();
    
    let mut x: i32 = 0;
    let mut y: i32 = 0;
    let mut pos: Vec<(i32,i32)> = Vec::new();
    let mut indices: BTreeMap<(i32,i32), Vec<usize>> = BTreeMap::new();
    indices.entry((x, y)).or_insert(vec![0]);
    pos.push((x, y));
    for i in 0..n {
        if b[i] == b'U' {
            y += 1;
        } else if b[i] == b'D' {
            y -= 1;
        } else if b[i] == b'L' {
            x -= 1;
        } else if b[i] == b'R' {
            x += 1;
        }
        let idx = indices.entry((x, y)).or_insert(Vec::new());
        idx.push(i+1);
        pos.push((x, y));
    }

    let query_left = |x: i32, y: i32, l: usize| -> bool {
        let key = (x, y);
        if !indices.contains_key(&key) {
            return false;
        }
        let idx: &Vec<usize> = indices.get(&key).unwrap();
        idx[0] <= l-1
    };

    let query_right = |x: i32, y: i32, r: usize| -> bool {
        let key = (x, y);
        if !indices.contains_key(&key) {
            return false
        }
        let idx: &Vec<usize> = indices.get(&key).unwrap();
        idx[idx.len()-1] >= r+1
    };

    let query_middle = |xi: i32, yi: i32, left: usize, right: usize| -> bool {
        let x = pos[left-1].0 + pos[right].0 - xi;
        let y = pos[left-1].1 + pos[right].1 - yi;
        let key = (x, y);
        if !indices.contains_key(&key) {
            return false;
        }
        let idx: &Vec<usize> = indices.get(&key).unwrap();
        let mut l: i32 = 0;
        let mut r: i32 = (idx.len()-1) as i32;
        let mut x: i32 = -1;
        while l <= r {
            let mid = l + (r-l) / 2;
            if idx[mid as usize] >= (left-1) {
                x = idx[mid as usize] as i32;
                r = mid-1;
            } else {
                l = mid+1;
            }
        }
        if x < 0 {
            return false;
        }
        x <= right as i32
    };

    let do_query = |xi: i32, yi: i32, l: usize, r: usize| -> bool {
        query_left(xi, yi, l) || query_right(xi, yi, r) || query_middle(xi, yi, l, r)
    };

    for _ in 0..q {
        let query = read_vec::<i32>();
        let result = do_query(query[0], query[1], query[2] as usize, query[3] as usize);
        if result {
            writeln!(writer, "YES").ok();
        } else {
            writeln!(writer, "NO").ok();
        }
    }
}

fn main() {
    solve();
}
